### PR TITLE
feat(server): repo-scoped skills overlay (#3067)

### LINKED ISSUE #3067: feat(server): repo-scoped skills overlay (./.chroxy/skills/) on top of global ~/.chroxy/skills/
## Context

The skills MVP (#2957) hardcodes `~/.chroxy/skills/` as the only source of skill files (`packages/server/src/skills-loader.js:17`). This is global per machine — the same skills are injected into every session regardless of which repo Claude is working in.

## Why this matters

Different projects often have different conventions:
- chroxy: ES modules + no semicolons + Zero Attribution Policy
- a personal blog: completely different style + tone
- a work repo: house style guide + security constraints

Today a user has to either:
1. Stuff every project's conventions into one global skills file (verbose, mostly irrelevant injection per session)
2. Manually swap the contents of `~/.chroxy/skills/` when changing projects (terrible UX)

## Proposed design

Two-tier overlay, walked at session start:

1. **Global skills** — `~/.chroxy/skills/*.md` (existing behavior)
2. **Repo-scoped skills** — `<session.cwd>/.chroxy/skills/*.md`

Both sets are concatenated and injected via the same per-provider mechanism (`systemPrompt.append` for Claude SDK, `--append-system-prompt` for Claude CLI, prepended to first user message for Codex/Gemini via `MessageTransformPipeline`).

**Conflict resolution:** if a repo-scoped file shares a filename with a global one, repo wins (treat as override). Filename-based dedup keeps the rule simple.

**Scope discovery:** walk up from `session.cwd` to find the nearest `.chroxy/skills/` (like `.git`). Optional — pure `cwd` is also acceptable for v1 of this feature.

## Acceptance criteria

- [ ] `loadActiveSkills()` accepts a `cwd` and merges global + repo skills
- [ ] Repo file overrides global file with same name
- [ ] WS `skills_available` payload labels each skill's source (`global` vs `repo`)
- [ ] Tests cover: global only, repo only, both with no conflict, both with conflict
- [ ] Doc update in `docs/skills.md`

## Why deferred

User-discovered during dogfooding the v1 MVP. Reasonable v2 territory — should land alongside or after #2958 (frontmatter) so the `providers:` and `activation:` fields can be specified per-repo too.

## References

- MVP: #2957 (closed 2026-04-24)
- Related v2 work: #2958 (frontmatter), #2959 (trust model)
- Loader: `packages/server/src/skills-loader.js`
- Doc: `docs/skills.md`

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

- Layer per-repo `.chroxy/skills/` on top of the global `~/.chroxy/skills/`. The repo overlay is discovered at session start by walking up from the session's cwd (same lookup pattern as `.git`).
- Repo skills override global skills on filename collision — supply a project-specific `coding-style.md` without losing other global skills.
- WS `skills_list` payload tags each entry with `source: "global" | "repo"` so clients can show which tier a skill came from.

## Why

The skills MVP (#2957) hardcoded a single global directory. As soon as you have more than one project, "stuff every convention into one global file" or "swap the directory contents between sessions" both feel bad. Discovering a per-repo overlay fixes the ergonomic gap with no behaviour change for users who only use the global tier.

## Implementation notes

- `skills-loader.js` — adds `loadActiveSkillsLayered({globalDir, repoDir})` and `findRepoSkillsDir(cwd)`. The original `loadActiveSkills(dir)` keeps its v1 signature; the new optional `{ source }` arg lets the layered loader annotate each skill.
- `BaseSession` — resolves the repo dir from `session.cwd` via walk-up. Tests can pin both layers with `skillsDir` + `repoSkillsDir` options. All four session subclasses (cli, sdk, codex, gemini) thread `repoSkillsDir` through their constructors.
- `handleListSkills` — derives `repoDir` from the active session's cwd. With no active session, returns global only.
- Protocol schema — `ServerSkillsListSchema.skills[].source` is `z.enum(['global', 'repo']).optional()`. Optional so v1 clients still parse pre-#3067 payloads cleanly.
- Test hygiene drive-by — `base-session.test.js` now pins `skillsDir`/`repoSkillsDir` to empty temp dirs. Previously it was reading the developer's real `~/.chroxy/skills/` and breaking on machines that had any skill installed (the same P1 anti-pattern called out for `settings.json`).

## Test plan

- [x] `skills-loader.test.js` — 8 new test cases cover global only, repo only, no overlap, filename override, partial override, walk-up discovery, the `globalDir === repoDir` edge case, and the source-tagging behaviour
- [x] `settings-handlers.test.js` — `list_skills` test cases cover repo-tier discovery from a session cwd, walk-up from a nested cwd, and the no-session fallback
- [x] `skills-integration.test.js` — existing end-to-end tests still pass (Cli, Codex, Gemini)
- [x] Server lint — no new warnings on touched files
- [x] App + dashboard `tsc --noEmit` — clean
- [ ] Try a per-repo skill in a real session (manual smoke after merge)

Closes #3067

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #3087 — repo-scoped skills overlay

### Summary
Layers a per-repo `.chroxy/skills/` overlay on top of the existing global `~/.chroxy/skills/`, discovered by walking up from the session's cwd (same pattern as `.git`). Repo files override global on filename collision, and the WS `skills_list` payload now tags each entry with `source: "global" | "repo"`. CI is green across all 12 checks.

This is a clean, well-scoped feature PR. The new code is straightforward, well-commented, and the tests are honest about what they can and can't cover (the `list_skills` repo-tier tests explicitly note they can't fake the user's real `~/.chroxy/skills`).

### Strengths
- **`loadActiveSkillsLayered` dedup is correct.** The `Map.set` ordering trick (globals first, then repos) gives repo-wins-on-collision in a single readable pass — much nicer than a manual filter. Test coverage hits the four key cases (global only, repo only, no overlap, partial overlap) plus the `globalDir === repoDir` edge case.
- **`source` field on the protocol schema is properly optional** (`z.enum([...]).optional()`), so v1 clients parsing pre-#3067 payloads don't break. Good backward-compat hygiene, and the inline comment in `server.ts` calls it out for future readers.
- **Defensive input handling on `findRepoSkillsDir`.** Null/undefined/non-string cwd, non-existent ancestors, and the file-vs-directory case (`.chroxy/skills` is a regular file) are all explicitly tested. The `dir !== prev` termination correctly handles the filesystem root where `dirname('/') === '/'`.
- **The `base-session.test.js` fix is a real bug, not a cosmetic touch.** Before this PR, that test constructed `BaseSession` with no `skillsDir` override, which made it read the developer's real `~/.chroxy/skills/` — same P1 anti-pattern the project flagged for `settings.json`. The PR description correctly identifies it as a drive-by, but the fix is genuinely load-bearing for test isolation.
- **Constructor plumbing is mechanical and consistent.** All four subclasses (`cli-session`, `sdk-session`, `codex-session`, `gemini-session`, plus the `jsonl-subprocess-session` intermediate) thread `repoSkillsDir` identically. No subtle differences to audit later.
- **Docs are updated** (`docs/skills.md` covers the new tier, the override rule, and the recommendation to commit repo skills alongside the project).

### Issues Found

**Critical:** None.

**Major:** None.

**Minor**

1. **`handleListSkills` recomputes the repo dir on every call.** The active session already discovered and cached `_repoSkillsDir` in its constructor. The handler does its own `findRepoSkillsDir(entry.session.cwd)` walk again on every `list_skills` request — extra `statSync` calls, and (more importantly) it diverges from the value the session is actually using if the repo dir was injected via `repoSkillsDir: null`/explicit override (e.g. in tests, or any future caller that wants to disable the overlay). Consider exposing `session._repoSkillsDir` (or a `getRepoSkillsDir()` accessor) and reusing it in the handler. Not blocking — the values will agree in production today — but it's a footgun the next person to touch this won't expect.

2. **`loadActiveSkillsLayered` re-scans the global dir twice when `globalDir === repoDir`.** Look at the `repos` ternary: when `sameDir` is true it calls `loadActiveSkills(globalDir, { source: 'repo' })`. That's fine, but the code path is a little awkward to follow. A two-line refactor — compute `effectiveGlobal` and `effectiveRepo` at the top, then load each — would make the dedup intent clearer. Cosmetic.

3. **`findRepoSkillsDir` does not stop at the user's home directory.** A session started in `~/scratch/foo/bar` (no `.git`, no `.chroxy/skills`) walks all the way up through `~`, `~/..`, `/Users`, `/`. If the user happens to have a `~/.chroxy/skills/` (which they do, by design — it's the global tier), the walk would actually match the global dir as a "repo" overlay! That triggers the `globalDir === repoDir` branch and tags every global skill as `source: 'repo'` for that session. Same data, but the source label silently lies for any user whose session cwd is anywhere under `$HOME` and not inside an actual repo. Probably worth either (a) stopping the walk at `homedir()` or at the first directory containing a `.git`, or (b) skipping the candidate if it equals `DEFAULT_SKILLS_DIR`. Filed as follow-up below since it's behavioural rather than a regression — v1 didn't have this concept at all.

**Nits**

- `REPO_DISCOVERY_MAX_DEPTH = 100` is fine in practice (real paths max out at ~20-30 levels); the cap is a sensible belt for pathological symlinks. No DoS concern — `statSync` on each level is cheap and the loop is bounded.
- The loader runs once at session construction. A skill added mid-session (`echo > ~/.chroxy/skills/new.md`) won't show up until session restart. This is consistent with v1 behaviour and the PR doesn't claim otherwise — out of scope.

### Deferred Items

| Finding | Issue |
|---|---|
| `findRepoSkillsDir` may match `~/.chroxy/skills` as a "repo" overlay when walking up from a cwd under `$HOME` outside any repo | #3088 |

### Architecture Notes

The two-tier design is the right shape — same pattern as `.gitignore` + `~/.config/git/ignore`, or `package.json` + `~/.npmrc`. The `source` tag in the WS payload sets up nicely for a future dashboard UI to show "Global" / "This repo" badges on each skill in the list. No protocol breakage; clients that ignore the new field continue to work.

The repo-scoped overlay also dovetails with the deferred v2 frontmatter work (#2958) — once skills carry `providers:` / `activation:` metadata, those can be specified per-repo too without further loader plumbing.

### Verdict

**Approve.** Ship it. The minor recompute concern (#1 above) and the handler/session inconsistency window it opens are worth a follow-up pass but don't block — the values will be identical in every production code path today. The home-dir walk-up edge case is filed as #3088 for a polish pass.

File: packages/server/tests/base-session.test.js [hunk 1]
```diff
@@ -1,12 +1,31 @@
...
 
 describe('BaseSession', () => {
   let session
+  let emptySkillsDir
 
   beforeEach(() => {
-    session = new BaseSession({ cwd: '/tmp', model: 'test-model', permissionMode: 'approve' })
+    // Pin skillsDir + repoSkillsDir to empty temp dirs so the tests don't
+    // pick up whatever lives in the developer's real ~/.chroxy/skills/ (#3067).
+    // cwd: '/tmp' is also passed an empty repoSkillsDir to bypass walk-up.
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/tests/base-session.test.js]: **FIX**

Fixed in `a63a9b065`

**Change:** Aligned the comment with the actual constructor argument.

```diff
-    // cwd: '/tmp' is also passed an empty repoSkillsDir to bypass walk-up.
+    // cwd: '/tmp' is also passed repoSkillsDir: null to bypass walk-up.
```

File: packages/server/src/handlers/settings-handlers.js [hunk 2]
```diff
@@ -364,16 +364,29 @@ function handleListProviders(ws, client, msg, ctx) {
...
+  const entry = resolveSession(ctx, msg, client)
+  const repoDir = entry?.session?.cwd ? findRepoSkillsDir(entry.session.cwd) : null
+  const skills = loadActiveSkillsLayered({
+    globalDir: DEFAULT_SKILLS_DIR,
+    repoDir,
+  }).map((s) => ({
     name: s.name,
     description: s.description,
+    source: s.source,
   }))
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/handlers/settings-handlers.js]: **FIX**

Fixed in `a63a9b065`

**Change:** `handleListSkills` now reads from `session._getSkills()` when an active session is bound, falling back to a fresh `loadActiveSkillsLayered` scan only when no session resolves (or it doesn't expose the accessor — e.g. mock sessions in tests). Keeps the WS payload aligned with what the session actually injected at construction and respects test-only `skillsDir`/`repoSkillsDir` overrides.

```diff
 function handleListSkills(ws, client, msg, ctx) {
   const entry = resolveSession(ctx, msg, client)
-  const repoDir = entry?.session?.cwd ? findRepoSkillsDir(entry.session.cwd) : null
-  const skills = loadActiveSkillsLayered({
-    globalDir: DEFAULT_SKILLS_DIR,
-    repoDir,
-  }).map((s) => ({
-    name: s.name,
-    description: s.description,
-    source: s.source,
-  }))
+
+  let skills
+  if (entry?.session && typeof entry.session._getSkills === 'function') {
+    const sessionSkills = entry.session._getSkills()
+    if (Array.isArray(sessionSkills)) {
+      skills = sessionSkills.map((s) => ({
+        name: s.name,
+        description: s.description,
+        source: s.source || 'global',
+      }))
+    }
+  }
+
+  if (!skills) {
+    const repoDir = entry?.session?.cwd ? findRepoSkillsDir(entry.session.cwd) : null
+    skills = loadActiveSkillsLayered({
+      globalDir: DEFAULT_SKILLS_DIR,
+      repoDir,
+    }).map((s) => ({
+      name: s.name,
+      description: s.description,
+      source: s.source,
+    }))
+  }
+
   ctx.send(ws, { type: 'skills_list', skills })
 }
```

The existing `list_skills` tests still pass — the mock session created via `createMockSession()` has no `_getSkills`, so it correctly falls through to the fresh-scan branch and exercises the cwd-derived repo lookup.

File: docs/skills.md [hunk 3]
```diff
@@ -4,21 +4,30 @@ Skills are reusable instruction snippets that Chroxy injects into every session
 
 ## Where skill files live
 
-Chroxy reads skills from `~/.chroxy/skills/`. Each skill is a plain Markdown file:
+Chroxy looks for skills in two places, then merges them at session start:
+
+1. **Global** — `~/.chroxy/skills/*.md` — applied to every session on this machine.
+2. **Repo overlay** — `<repo>/.chroxy/skills/*.md` — discovered by walking up from the session's working directory (same lookup pattern as `.git`). Applied only to sessions started inside that repo.
```
[the remark this answers is not in the document]
Comment (blamechris) [reply docs/skills.md:11]: **FIX**

Fixed in `a63a9b065`

**Change:** Added a Trust note to `docs/skills.md` warning that repo-overlay skills shape the system prompt of any session started inside that repo, so working in an untrusted clone is a prompt-injection / exfil-via-tool-call risk. Includes a quick opt-out (run from a cwd with no `.chroxy/skills/` ancestor, or rename the dir) and points at #2959 as the place where a first-class trust model with provenance + per-skill enable/disable will land.

```diff
 When a global file and a repo file share the same filename, the **repo file wins** — treat it as an override, not an addition. Filename-based dedup keeps the rule predictable.
+
+> **Trust note.** Repo-overlay skills are auto-loaded any time you start a session inside that repo's tree, and their content is injected directly into the model's system prompt. That means cloning or opening an *untrusted* repo can shape model behaviour — including potential prompt injection or guidance to exfiltrate data via tool calls. Treat `<repo>/.chroxy/skills/` like any other code in the repo: review it before working in an unfamiliar checkout, and avoid running Chroxy inside repos you don't trust. To opt out for a session, run Chroxy from a `cwd` that has no `.chroxy/skills/` in any ancestor (or temporarily rename the directory). A first-class trust model with provenance and per-skill enable/disable is tracked in #2959.
```

**Comment by blamechris** [comment]
## Review Comments Addressed

| # | Comment | Outcome | Reference |
|---|---------|---------|-----------|
| 1 | Comment text in `base-session.test.js` said "empty repoSkillsDir" but the constructor is given `repoSkillsDir: null` | FIX | `a63a9b065` |
| 2 | `handleListSkills` re-scans disk every call instead of mirroring the session's already-loaded skills (diverges from test overrides; can drift mid-session) | FIX | `a63a9b065` |
| 3 | `docs/skills.md` should warn that repo overlays from untrusted clones can shape the system prompt (prompt-injection / exfil risk) | FIX | `a63a9b065` |

**Total:** 3 comments addressed
- Fixed: 3 (commit `a63a9b065`)
- False positives: 0
- Follow-up issues created: 0
- Existing issues closed: 0
